server/service/osquery.go

```diff
@@ -1405,3 +1414,171 @@ func (svc *Service) SubmitResultLogs(ctx context.Context, logs []json.RawMessage
+	}
+	if appConfig.ServerSettings.QueryReportsDisabled {
+		return nil
+	}
+
+	var queryResults []fleet.ScheduledQueryResult
+	for _, raw := range results {
+		var result fleet.ScheduledQueryResult
+		if err := json.Unmarshal(raw, &result); err != nil {
+			return newOsqueryError("unmarshalling individual result log: " + err.Error())
+		}
+		queryResults = append(queryResults, result)
+	}
+
+	// Filter results to only the most recent for each query
+	filtered := getMostRecentResults(queryResults)
+
+	for _, result := range filtered {
+		if err := svc.processResults(ctx, result); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (svc *Service) processResults(ctx context.Context, result fleet.ScheduledQueryResult) error {
+	// Discard result if there is no snapshot
+	if len(result.Snapshot) == 0 {
+		return nil
+	}
+
+	teamID, queryName, err := getQueryNameAndTeamIDFromResult(result.QueryName)
+	if err != nil {
+		return newOsqueryError("getting query name and team ID: " + err.Error())
+	}
+
+	query, err := svc.ds.QueryByName(ctx, teamID, queryName)
+	if err != nil {
+		return newOsqueryError("getting query by Name: " + err.Error())
+	}
+
+	// Discard Result if query is marked as discard data or if the query report is full
+	rowCount, err := svc.ds.ResultCountForQuery(ctx, query.ID)
+	if err != nil {
+		return newOsqueryError("getting result count for query: " + err.Error())
+	}
+	if query.DiscardData || rowCount >= fleet.MaxQueryReportRows {
+		return nil
+	}
+
+	host, err := svc.ds.HostByIdentifier(ctx, result.OsqueryHostID)
+	if err != nil {
+		return newOsqueryError("getting host ID: " + err.Error())
+	}
+
+	return svc.overwriteResultRows(ctx, result, query.ID, host.ID, rowCount)
+}
+
+// The "snapshot" array in a ScheduledQueryResult can contain multiple rows.  Each
+// row is saved as a separate ScheduledQueryResultRow. ie. a result could contain
+// many USB Devices or a result could contain all User Accounts on a host.
+func (svc *Service) overwriteResultRows(ctx context.Context, result fleet.ScheduledQueryResult, queryID, hostID uint, rowCount int) error {
+	fetchTime := time.Now()
+
+	// Update row count to account for deletion of existing host results
+	hostResultCount, err := svc.ds.ResultCountForQueryAndHost(ctx, queryID, hostID)
+	if err != nil {
+		return newOsqueryError("getting result count for query and host: " + err.Error())
+	}
+	rowCount = rowCount - hostResultCount
+
+	rows := make([]*fleet.ScheduledQueryResultRow, 0, len(result.Snapshot))
+
+	for _, snapshotItem := range result.Snapshot {
+
+		row := &fleet.ScheduledQueryResultRow{
+			QueryID:     queryID,
+			HostID:      hostID,
+			Data:        snapshotItem,
+			LastFetched: fetchTime,
+		}
+
+		rows = append(rows, row)
+
+		// only insert enough rows to reach the maxQueryReportRows
+		rowCount++
+		if rowCount >= fleet.MaxQueryReportRows {
+			break
+		}
+	}
+
+	if err := svc.ds.OverwriteQueryResultRows(ctx, rows); err != nil {
```

Comment:
What happens if you have simultaneous hosts submitting their results? I believe there could be a race condition (Maybe we need to do `ResultCountForQueryAndHost` and `OverwriteQueryResultRows` on the same transaction?)

Reply:
agree, will implement